consul/autopilot.go

```diff
@@ -0,0 +1,286 @@
+}
+
+// serverHealthLoop monitors the health of the servers in the cluster
+func (s *Server) serverHealthLoop() {
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(s.config.ServerHealthInterval)
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			serverHealths := make(map[string]*structs.ServerHealth)
+
+			state := s.fsm.State()
+			_, autopilotConf, err := state.AutopilotConfig()
+			if err != nil {
+				s.logger.Printf("[ERR] consul: error retrieving autopilot config: %s", err)
+			}
+
+			// Build an updated map of server healths
+			for _, member := range s.LANMembers() {
+				if member.Status == serf.StatusLeft {
+					continue
+				}
+
+				valid, parts := agent.IsConsulServer(member)
+				if valid {
+					health := s.queryServerHealth(member, parts, autopilotConf)
+					serverHealths[parts.Addr.String()] = health
+				}
+			}
+
+			s.autopilotLock.Lock()
+			s.autopilotHealth = serverHealths
+			s.autopilotLock.Unlock()
+
+			if err := s.promoteNonVoters(autopilotConf); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for non-voters to promote: %s", err)
+			}
+		}
+	}
+}
+
+// removeDeadLoop checks for dead servers periodically, or when receiving on autopilotRemoveDeadCh
+func (s *Server) removeDeadLoop() {
+	ticker := time.NewTicker(s.config.RemoveDeadInterval)
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		case <-s.autopilotRemoveDeadCh:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// pruneDeadServers removes up to numPeers/2 failed servers
+func (s *Server) pruneDeadServers() error {
+	state := s.fsm.State()
+	_, autopilotConf, err := state.AutopilotConfig()
+	if err != nil {
+		return err
+	}
+
+	// Look for dead servers to clean up
+	peers, err := s.numPeers()
+	if err != nil {
+		return err
+	}
+
+	removed := 0
+	if autopilotConf.CleanupDeadServers {
+		for _, member := range s.serfLAN.Members() {
+			// Exit early if we already removed the max amount of servers
+			if removed == peers/2 {
+				break
+			}
+
+			valid, _ := agent.IsConsulServer(member)
+			if valid && member.Status == serf.StatusFailed {
+				removed++
+				s.logger.Printf("[INFO] consul: Attempting removal of failed server: %v", member.Name)
+				go s.serfLAN.RemoveFailedNode(member.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// promoteNonVoters promotes eligible non-voting servers to voters.
+func (s *Server) promoteNonVoters(autopilotConf *structs.AutopilotConfig) error {
+	minRaftProtocol, err := ServerMinRaftProtocol(s.LANMembers())
+	if err != nil {
+		return fmt.Errorf("error getting server raft protocol versions: %s", err)
+	}
+
+	if minRaftProtocol >= 3 {
```

Comment:
You could skip out if the min `< 3` to de-nest all this. Might be a good place to print a DEBUG message as well so it's easy to tell why this isn't kicking in.